relayer/src/link.rs

```diff
@@ -411,45 +442,95 @@ impl RelayPath {
-                .decrement()
-                .map_err(|e| LinkError::Failed(e.to_string()))?;
-
-            self.relay_pending_packets()?;
+            self.relay_pending_packets(height)?;
             self.clear_packets = false;
         }
 
         Ok(())
     }
 
     /// Iterate through the IBC Events, build the message for each and collect all at same height.
-    /// Send a multi message transaction with these, prepending the client update
     pub fn relay_from_events(&mut self, batch: EventBatch) -> Result<(), LinkError> {
         self.clear_packets(batch.height)?;
 
-        // collect relevant events in self.all_events
+        // Collect relevant events in `self.all_events`.
         self.collect_events(&batch.events);
         self.adjust_events_height()?;
 
+        if self.channel.connection_delay.as_nanos() > 0 {
+            self.relay_with_delay()
+        } else {
+            self.relay_multi_msg()
+        }
+    }
+
+    /// Sends the client update in a separate transaction from packet messages, allowing for the
+    /// delay period to pass between the client update and packet messages.
+    fn relay_with_delay(&mut self) -> Result<(), LinkError> {
+        // If there are some new events, update src. and dest. clients, then schedule the relaying
+        // of packets corresponding to these events.
+        if !self.all_events.is_empty() {
+            let dst_update_height = self.dst_latest_height()?;
+            let src_update_height = self.all_events[0].height().increment();
+            self.update_clients(src_update_height, dst_update_height)?;
+
+            // Schedule the event batch, and let it sit for a while until the delay period passes
+            self.schedule_batch(self.all_events.clone(), dst_update_height);
+        }
+
+        // Iterate through all the outstanding event batches that have fulfilled their delay period,
+        // and send the packet messages corresponding to these events
+        while let Some(batch) = self.try_fetch_scheduled_batch() {
+            self.all_events = batch.events;
+
+            for _i in 0..MAX_ITER {
+                self.reset_buffers();
+
+                // Collect the messages for all events
+                for event in self.all_events.clone() {
+                    debug!("{} => {:?}", self.src_chain.id(), event);
+                    self.handle_packet_event(&event, batch.dst_height)?;
+                }
+
+                // Send messages
+                let res = self.send_msgs()?;
+                debug!("\nresult {:?}", res);
+
+                if self.all_events.is_empty() {
+                    break;
+                }
+
+                debug!("retrying");
+            }
+        }
+
+        self.all_events = vec![];
+        Ok(())
+    }
+
+    /// Send a multi message transaction with packet messages, prepending the client update.
+    /// Ignores any delay period for packets.
+    fn relay_multi_msg(&mut self) -> Result<(), LinkError> {
```

Comment:
They are all multi messages, maybe:
```suggestion
    fn relay_immediately(&mut self) -> Result<(), LinkError> {
```